### PR TITLE
Unpin `numpy`

### PR DESCRIPTION
**Pull request by mudit2812**
## Context:
Previously, PennyLane pinned NumPy to version `numpy<2.1` (see the [first commit](https://github.com/PennyLaneAI/pennylane/pull/7039/commits/449ae04f3fb308fdd7b92dacadd972f1a08390a0) of this PR). This restriction created compatibility challenges for users building their own projects with PennyLane as a dependency, as reported in #7033. 

We've determined there are no fundamental blockers preventing PennyLane from supporting newer NumPy versions - only historical integration issues, particularly with TensorFlow, that have caused cascading CI problems and potential ecosystem compatibility concerns.

With NumPy 1.26's end-of-support approaching in September 2025 ([reference](https://endoflife.date/numpy)), it's critical we thoroughly investigate and address these obstacles. This PR aims to explore unpinning NumPy, document issues encountered, implement fixes, and adjust our CI pipeline accordingly.
Screenshot from 2025-03-18 10-37-25

Besides, we can check the NEP 29 Drop Schedule: https://numpy.org/neps/nep-0029-deprecation_policy.html

## Description of the Change:
This PR follows a systematic approach:

1. Initially unpinning NumPy~1.26 but setting `>= 2.0.1` to ensure the main environment uses NumPy 2
2. Documenting and analyzing all issues encountered with NumPy 2 compatibility
3. Methodically resolving each identified issue
4. Fully unpinning NumPy to allow maximum flexibility
5. Updating CI workflows as needed to accommodate version differences
6. Adding demonstrations showing PennyLane works safely with projects requiring NumPy 2

Specific changes include:
- [x] Updated `tensorflow` to version `2.18` for NumPy 2 compatibility across multiple files: 
  - `.github/workflows/interface-dependency-versions.yml`
  - `docker/interfaces/install-interface-gpu.sh`
  - `docker/interfaces/install-interface.sh`
  - `requirements.txt`
  - `doc/requirements.txt`
  - `setup.py`
- [x] Updated `jax` and `jaxlib` to `0.4.28` for NumPy 2 compatibility in `doc/requirements.txt`
- [x] Pinned `ml_dtypes` for compatibility in `requirements.txt` and `setup.py`
- [x] Added conditional xfail decorator for `KerasLayer` tests that won't pass with NumPy 2 (pending deprecation)
- [x] Added conditional xfail for `qchem` tests requiring `openfermion` based on NumPy version

Extra manual test:
 - [x] Locally installation of `numpy-2.2` works without conflicts with PL
 - [x] Ci tests with other affected packages in PL ecosys passed:
 - [x] lightning: https://github.com/PennyLaneAI/pennylane-lightning/pull/1095
 - [x] pennylane-cirq: https://github.com/PennyLaneAI/pennylane-cirq/pull/211
 - [x] pennylane-qiskit: https://github.com/PennyLaneAI/pennylane-qiskit/pull/622
 

## Benefits:
- Identified and resolved root causes of NumPy 2.x CI failures
- Users can now use PennyLane with NumPy 2.x environments without compatibility issues
- Improved future maintainability by removing artificial version constraints
- Better aligned with the Python ecosystem's forward progress

## Remaining Challenges:
We encountered several issues that warrant further investigation, though not crucially important for now:
- Sphinx documentation generation succeeded with NumPy 1.x but fails with NumPy 2.x despite no obvious connection, since the documentation building should have their own dependencies in principle, as `doc/requirements.txt` exists. However, this was reflected possibly resulting from that there used to be some spaghetti between main requiremnts and doc/requirements.
- Inconsistent test results between CI environment and local development environment when using NumPy 2 + TensorFlow 2.18 + Keras 3. Basically, if one tested Keras on their local machine, they should see ~half~ one third of them to fail; however in our CI, actually only one of them fails; this is also why the xfail macro was defined by not widely used in the affected test file. This phenomenon could be caused by some special environment variables in workflow setup, s.t. eventually the actual Keras used is still Keras2.

## Related GitHub Issues:
[sc-85628]

### CODE REVIEW

**Comment by JerryChen97** [comment]
No need for full test stuies for now since we should fix the issue with incompatibility ml dtype first

File: doc/requirements.txt [hunk 1]
```diff
@@ -21,7 +22,7 @@ sphinxcontrib-devhelp==1.0.2
 sphinxcontrib-qthelp==1.0.3
 sphinxcontrib-htmlhelp==2.0.1
 sphinxcontrib-serializinghtml==1.1.5
-tensorflow==2.11.1; platform_machine == "x86_64"
+tensorflow==2.19.0; platform_machine == "x86_64"
```
Comment (JerryChen97) [inline doc/requirements.txt:25]:
```suggestion
tensorflow==2.18.0; platform_machine == "x86_64"
```
`2.19` is not necessary. Let's pin consistent with other places

**Comment by JerryChen97** [comment]
We also need an extra workflow to add to `.github` so that we test all the `tf` marked test suites with `numpy>=2.1` to reveal any potential conflicts in the future. Would like to hear any suggestsions or comments from @rashidnhm

File: pennylane/qnn/keras.py [hunk 2]
```diff
@@ -352,7 +347,10 @@ def __init__(
 
         self.qnode_weights = {}
 
-        super().__init__(dynamic=True, **kwargs)
+        if CORRECT_KERAS_VERSION:
+            super().__init__(dynamic=True, **kwargs)
+        else:
```
Comment (JerryChen97) [inline pennylane/qnn/keras.py:352]:
```suggestion
        else: # pragma: no cover
```

File: pennylane/qnn/keras.py [hunk 3]
```diff
@@ -13,6 +13,7 @@
 # limitations under the License.
 """This module contains the classes and functions for integrating QNodes with the Keras Layer
 API."""
+# pragma: no cover
```
Comment (JerryChen97) [inline pennylane/qnn/keras.py:16]:
```suggestion
```

**Comment by JerryChen97** [comment]
We are pretty confident with current solution after several rounds of full tests. Now rm the label for faster checks and less CI consumption

**Comment by rashidnhm** [comment]
> We also need an extra workflow to add to `.github` so that we test all the `tf` marked test suites with `numpy>=2.1` to reveal any potential conflicts in the future. Would like to hear any suggestsions or comments from @rashidnhm

What kinds of conflicts do we expect? Dependency conflict or straight up functionality breaking?

**Comment by JerryChen97** [comment]
> > We also need an extra workflow to add to `.github` so that we test all the `tf` marked test suites with `numpy>=2.1` to reveal any potential conflicts in the future. Would like to hear any suggestsions or comments from @rashidnhm
> 
> What kinds of conflicts do we expect? Dependency conflict or straight up functionality breaking?

Functionality breaking.

File: tests/qchem/openfermion_pyscf_tests/test_molecular_hamiltonian.py [hunk 4]
```diff
@@ -958,7 +959,7 @@ def test_error_raised_for_incompatible_type():
 
 def test_error_raised_for_missing_molecule_information():
     r"""Test that molecular_hamiltonian raises an error when symbols, and coordinates
-    information is not provided.
+    informtest_coordinate_units_for_molecular_hamiltonianation is not provided.
```
Comment (andrijapau) [inline tests/qchem/openfermion_pyscf_tests/test_molecular_hamiltonian.py:962]: what happened here? 😆

**Review by andrijapau — COMMENTED** [review]
Initial pass through, everything makes sense to me. Why are those PRs tagged in the description closed?

**Comment by JerryChen97** [comment]
> Initial pass through, everything makes sense to me. Why are those PRs tagged in the description closed?

'cause they were only setup for the compatibility test. I closed them since tests passed without obvious errors. Didn't del the branches in case we need to tweak something later to test again.

File: tests/qchem/openfermion_pyscf_tests/test_molecular_hamiltonian.py (hunk 4 above)
Comment (JerryChen97) [reply tests/qchem/openfermion_pyscf_tests/test_molecular_hamiltonian.py:962]: Lollll I'll revert

File: doc/requirements.txt [hunk 5]
```diff
@@ -2,8 +2,9 @@ pip
 appdirs
 autograd
 autoray
-jax==0.4.16
-jaxlib==0.4.16
+cachetools
```
Comment (mlxd) [inline doc/requirements.txt:5]: The reason we need this is that PennyLane isn't installed by default in the sphinx env, so we must explicitly set the dependence here?

File: requirements.txt [hunk 6]
```diff
@@ -1,4 +1,5 @@
-numpy~=1.23
+numpy>=1.26
```
Comment (mlxd) [inline requirements.txt:1]: Do we need to set the lower bound to 1.26 here?

File: requirements.txt [hunk 7]
```diff
@@ -1,4 +1,5 @@
...
+ml_dtypes>=0.4.0
```
Comment (mlxd) [inline requirements.txt:2]: Do we need to define this? Is it a soft/hard dependency?

File: tests/qchem/openfermion_pyscf_tests/conftest.py [hunk 8]
```diff
@@ -16,10 +16,25 @@
 """
 import shutil
 
+import numpy as np
 import pytest
+from packaging.version import Version
```
Comment (mlxd) [inline tests/qchem/openfermion_pyscf_tests/conftest.py:21]: Any reason we use `packaging` instead of `importlib` for this?

File: tests/qchem/openfermion_pyscf_tests/test_dipole_of.py [hunk 9]
```diff
@@ -17,6 +17,7 @@
 # pylint: disable=too-many-arguments
 import numpy as np
 import pytest
+from conftest import xfail_on_numpy2  # pylint: disable=no-name-in-module
```
Comment (mlxd) [inline tests/qchem/openfermion_pyscf_tests/test_dipole_of.py:20]: Do we need to import with this? I assumed the conftest definitions are implicitly available -- or is that only for fixtures?

File: tests/qnn/test_keras.py [hunk 10]
```diff
@@ -75,6 +98,14 @@ def model_dm(get_circuit_dm, n_qubits, output_dim):
     return m
 
 
+@pytest.fixture(autouse=True)
+def reset_keraslayer_input_arg():
```
Comment (mlxd) [inline tests/qnn/test_keras.py:102]: How does this work?

File: tests/qnn/test_keras.py [hunk 11]
```diff
@@ -826,20 +845,22 @@ def circuit(x, weights):
...
-    for x_, r in zip(x, res):
-        assert qml.math.allclose(r, circuit(x_, layer.qnode_weights["weights"]), atol=tol)
+        for x_, r in zip(x, res):
+            assert qml.math.allclose(r, circuit(x_, layer.qnode_weights["weights"]), atol=tol)
 
-    # reset back to the old name
-    KerasLayer.set_input_argument("inputs")
+    finally:
+        # reset back to the old name
+        KerasLayer.set_input_argument("inputs")
```
Comment (mlxd) [inline tests/qnn/test_keras.py:863]: If we defined that function earlier for reseting the Keras inputs, do we want to make it reusable here as a fixture, instead of explicitly defining similar functionality?

File: tests/qnn/test_keras.py [hunk 12]
```diff
@@ -979,6 +1002,7 @@ def circuit(inputs, w1, w2):
     assert info["device_name"] == "default.qubit"
 
 
+@requires_keras2
```
Comment (mlxd) [inline tests/qnn/test_keras.py:1005]: Is this the only place we use this decorator? If so, does it make sense to keep it if we only use it once?

File: doc/requirements.txt (hunk 5 above)
Comment (JerryChen97) [reply doc/requirements.txt:5]: Yes! Full log can be found: https://github.com/PennyLaneAI/pennylane/actions/runs/13838317452/job/38718984043

File: requirements.txt (hunk 6 above)
Comment (JerryChen97) [reply requirements.txt:1]: Yeah I set it here since @Alex-Preciado suggested about lower bound in a conversation before. Looking forwrad to further discussion here

File: tests/qchem/openfermion_pyscf_tests/conftest.py (hunk 8 above)
Comment (JerryChen97) [reply tests/qchem/openfermion_pyscf_tests/conftest.py:21]: No I don't have personal preference here. I used this only because I saw we used it other places in PL before.

File: tests/qchem/openfermion_pyscf_tests/test_dipole_of.py (hunk 9 above)
Comment (JerryChen97) [reply tests/qchem/openfermion_pyscf_tests/test_dipole_of.py:20]: `xfail_on_numpy2` has been defined as a decorator rather than fixture in `conftest.py` so it still needs to be imported

File: tests/qnn/test_keras.py (hunk 12 above)
Comment (JerryChen97) [reply tests/qnn/test_keras.py:1005]: This is the weirdest phenomenon I observed these days. Actually, in my local tests 1/3 of the test should fail. However, in our CI eventually only this test fails.
Since `KerasLayer` will be deprecated soon and removed in 0.42, I'll say totally fine to get rid of the decorator.

File: tests/qnn/test_keras.py (hunk 10 above)
Comment (JerryChen97) [reply tests/qnn/test_keras.py:102]: So this and the other place where `finally:` appeared are both due to the fact that `KerasLayer` has a method to modify its own class property `input_arg` (still don't understand why we need such method and testing for them though) so I would like to safeguard no other tests launched together are affected.

File: tests/qnn/test_keras.py (hunk 11 above)
Comment (JerryChen97) [reply tests/qnn/test_keras.py:863]: You are right this one should be safe to remove

Comment (JerryChen97) [reply tests/qnn/test_keras.py]: Trying here https://github.com/PennyLaneAI/pennylane/pull/7039/commits/4beefa1c80048b76862ad0ca64150b189e99c1af

File: tests/qnn/test_keras.py (hunk 12 above)
Comment (JerryChen97) [reply tests/qnn/test_keras.py]: https://github.com/PennyLaneAI/pennylane/pull/7039/commits/c4b8c6c352a4b60d48a7074728eeed00dcf3d384

File: tests/qnn/test_keras.py (hunk 11 above)
Comment (JerryChen97) [reply tests/qnn/test_keras.py]: CI passed

File: doc/requirements.txt (hunk 5 above)
Comment (mlxd) [reply doc/requirements.txt:5]: I may be wrong, but I think the issue here is the lack of a venv and/or a miconfiguration between the pyproject.toml file and setup.py
This dep should be pulled in from the install of PennyLane, and shouldn't need to be specified. I think we can ignore it here, but it may be worth creating an issue.

File: requirements.txt (hunk 6 above)
Comment (mlxd) [reply requirements.txt:1]: If we need this, we can set it, though I am wondering if 1.26 is too strict.

Comment (Alex-Preciado) [reply requirements.txt:1]: No need to set a lower bound unless it’s the only way to ensure compatibility with both 1.x and 2.x. Do we run into any issues if we stick with the original constraint (`numpy~=1.23`)?

Comment (JerryChen97) [inline requirements.txt:1]:
```suggestion
numpy~=1.23
```

File: setup.py [hunk 13]
```diff
@@ -21,7 +21,8 @@
     version = f.readlines()[-1].split()[-1].strip("\"'")
 
 requirements = [
-    "numpy<2.1",
+    "numpy>=1.26",
```
Comment (JerryChen97) [inline setup.py:24]:
```suggestion
    "numpy",
```

File: setup.py [hunk 14]
```diff
@@ -21,7 +21,8 @@
...
+    "ml_dtypes>=0.4.0",
```
Comment (JerryChen97) [inline setup.py:25]:
```suggestion
    "ml_dtypes",
```

File: requirements.txt (hunk 6 above)
Comment (Alex-Preciado) [reply requirements.txt]: Do we want to consider NumPy's Drop Schedule ([here](https://numpy.org/neps/nep-0029-deprecation_policy#drop-schedule))? (This table [here](https://endoflife.date/numpy) may also be useful).

Comment (JerryChen97) [reply requirements.txt]: It is good! I'll just leave them as was and only make the numpy change in `setup.py` from previous `<2.1` to no boundaries.

File: requirements.txt (hunk 7 above)
Comment (JerryChen97) [reply requirements.txt:2]: Just freed `ml_dtypes`

File: requirements.txt [hunk 15]
```diff
@@ -1,4 +1,5 @@
 numpy~=1.23
+ml_dtypes>=0.4.0
```
Comment (JerryChen97) [inline requirements.txt:2]:
```suggestion
```

File: setup.py [hunk 16]
```diff
@@ -21,7 +21,8 @@
     version = f.readlines()[-1].split()[-1].strip("\"'")
 
 requirements = [
-    "numpy<2.1",
+    "numpy",
+    "ml_dtypes",
```
Comment (JerryChen97) [inline setup.py:25]:
```suggestion
```

File: tests/qchem/openfermion_pyscf_tests/conftest.py (hunk 8 above)
Comment (mlxd) [reply tests/qchem/openfermion_pyscf_tests/conftest.py:21]: If packaging is already in the venv as a dep, that's fine. importlib would be relying completely on the Python standard lib though. If this paradigm is used elsewhere, we can keep it for now.

File: doc/requirements.txt (hunk 5 above)
Comment (mlxd) [reply doc/requirements.txt:5]: Once we have an issue tracked for this, I'll give a :+1:

**Comment by JerryChen97** [comment]
@mlxd @Alex-Preciado I also posted the numpy end-of-support plan in the description now. Since it's very close to the end of security support (June 2025), shall we re-consider a lower bound for numpy here?

**Comment by mlxd** [comment]
> @mlxd @Alex-Preciado I also posted the numpy end-of-support plan in the description now. Since it's very close to the end of security support for 1.25 (June 2025), shall we re-consider a lower bound for numpy here?

No need to consider a lower bound.

File: doc/requirements.txt (hunk 5 above)
Comment (JerryChen97) [reply doc/requirements.txt:5]: https://github.com/PennyLaneAI/pennylane/issues/7106
@mlxd Hopefully I put things clear there! Feel free to edit it

Comment (mlxd) [reply doc/requirements.txt:5]: Perfect, thanks.

**Review by mlxd — APPROVED** [review]

**Review by andrijapau — APPROVED** [review]
Makes sense to me. Thanks for the descriptive PR description. 🚀
